src/work/BasicWork.h

```diff
@@ -0,0 +1,114 @@
+    static size_t const RETRY_A_FEW;
+    static size_t const RETRY_A_LOT;
+    static size_t const RETRY_FOREVER;
+
+    enum State
+    {
+        WORK_RUNNING,
+        WORK_SUCCESS,
+        WORK_WAITING,
+        WORK_FAILURE_RETRY,
+        WORK_FAILURE_RAISE,
+        WORK_DESTRUCTING
+    };
+
+    BasicWork(Application& app, std::function<void()> callback,
+              std::string name, size_t maxRetries);
+    virtual ~BasicWork();
+
+    std::string const& getName() const;
+    virtual std::string getStatus() const;
+    State getState() const;
+    bool isDone() const;
+
+    size_t getMaxRetries() const;
+    uint64_t getRetryETA() const;
+
+    // Main method for state transition, mostly dictated by `onRun`
+    void crankWork();
+
+    // TODO (mlo) look into uses of reset in current interface
+    // Proper cleanup needs to happen at some point, and `reset`
+    // might not be the best place. For instance, `GetRemoteFileWork`
+    // calls std::remove is `onReset`, so an attempt to remove a non-existent
+    // file happens every time a child is added, which is inefficient.
+    void reset();
+
+  protected:
+    // Implementers can override these callbacks to customize functionality
+    // `onReset` is called on any important state transition, e.g. when new work
+    // is added, on a retry, success and failure. This method ensures that
+    // all the references (including children) are properly cleaned up for
+    // safe destruction. `onRun` performs necessary work logic, and hints the
+    // next state transition. `onSuccess` is called when work transitions into
+    // WORK_SUCCESS state. Similarly, `onFailure*` methods are called upon
+    // transitioning into appropriate failure state. An example usage would be
+    // cleaning up side effects, like downloaded files in these methods. If you
+    // want to force failure (and reset / retry) you can return
+    // WORK_FAILURE_RETRY or WORK_FAILURE_RAISE. After a retry count is
+    // passed, WORK_FAILURE_RETRY means WORK_FAILURE_RAISE anyways.
+    // WORK_FAILURE_FATAL is equivalent to WORK_FAILURE_RAISE passed up in
+    // the work chain - when WORK_FAILURE_FATAL is raised in one work item,
+    // all work items that leaded to this one will also fail without retrying.
+
+    virtual void onReset();
+    virtual State onRun() = 0;
+    virtual void onFailureRetry();
+    virtual void onFailureRaise();
+    virtual void onSuccess();
+    virtual void onWakeUp();
+
+    // A helper method that implementers can use if they plan to
+    // utilize WAITING state. This tells the work to return to RUNNING
+    // state, and propagate the notification up to the scheduler
+    // An example use of this would be RunCommandWork:
+    // a timer is used to async_wait for a process to exit,
+    // with a call to `wakeUp` upon completion.
+    void wakeUp();
+
+    // Default wakeUp callback that implementers can use
+    std::function<void()> wakeUpCallback();
```

Comment:
there is no need for this default call back like this: `std::function` can take a `nullptr` value

Reply:
This is a helper function that implementers can use if they want a basic wakeUp callback (e.g., in `Work::addWork`)